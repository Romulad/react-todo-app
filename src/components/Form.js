import React, {useState} from "react";

export const errorsObj = {
    "Invalid name" : "Enter a valid task name",
    "Same name" : "Task with this name already exist"
};

export const checkSameTask = (name, tasks) => {
    let matchedTask = tasks?.filter(task => task.name.toLowerCase() === name.toLowerCase());
    return matchedTask.length > 0;

};

export const checkIput = (name, tasks) =>{
    let errorNames = Object.keys(errorsObj);
    if(name.length <= 1){
        return [false, errorNames[0]];
    }else if(checkSameTask(name, tasks)){
        return [false, errorNames[1]];
    }else{
        return [true, ''];
    };
};

export function ErrorContainer(props) {
    return (
        <p className="text-danger" style={{marginTop : 10}}>
            {props.text}
        </p>
    );
};


function Form(props){
    const [name, setName] = useState('');
    const [error, setError] = useState("");

    function handleSubmit(e){
        e.preventDefault();
        let inputIsValid = checkIput(name, props.tasks);
        if(!inputIsValid[0]){
            setError(inputIsValid[1]);
            return;
        }else{
            setError("");
        }
        props.addTask(name);
        setName('');
    };

    function handleDclick(e){
        e.preventDefault();
        props.setTask([]);
        localStorage.clear();
    };

    let deleDisabled = (
        <button 
            disabled
            className="btn btn__danger"
            style={{cursor : "not-allowed"}}>
            Delete all tasks
        </button> 
    );

    let deleteBtn = (
        <button 
            onClick={handleDclick}
            className="btn btn__danger">
            Delete all tasks
        </button> 
    );

    return(
        <form onSubmit={handleSubmit}>
            <h2 className="label-wrapper">
            <label 
                htmlFor="new-todo-input" 
                className="label__lg" 
                style={{
                    marginBottom : 20,
                    fontSize : 20
                }}>
                What needs to be done today?
            </label>
            </h2>
            <input
                type="text"
                id="new-todo-input"
                className="input input__lg"
                name="text"
                autoComplete="off"
                value={name}
                onChange={(e) => {setName(e.target.value)}}
                style={{borderRadius:"0.25rem"}}
            />
            {/* Show form validation error */}
            {error ? <ErrorContainer text={errorsObj[error]} /> : null} 
            <div className="btn-group">
                <button 
                    type="submit" 
                    className="btn btn__primary">
                    New task
                </button>
                {props.tasks.length <= 0 ? deleDisabled : deleteBtn}
            </div>
        </form>
    )
};

export default Form;