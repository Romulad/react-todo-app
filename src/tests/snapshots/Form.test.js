import renderer from "react-test-renderer"

import Form, {ErrorContainer} from "../../components/Form";


test("Form component do not change", ()=>{
    const component = renderer.create(
        <Form tasks={[{name:"lunch"}]}></Form>
    );
    
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("ErrorContainer do not change", ()=>{
    const component = renderer.create(
        <ErrorContainer text='An error'></ErrorContainer>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})