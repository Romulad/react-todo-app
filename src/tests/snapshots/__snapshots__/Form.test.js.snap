// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorContainer do not change 1`] = `
<p
  className="text-danger"
  style={
    Object {
      "marginTop": 10,
    }
  }
>
  An error
</p>
`;

exports[`Form component do not change 1`] = `
<form
  onSubmit={[Function]}
  style={
    Object {
      "marginTop": 10,
    }
  }
>
  <h2
    className="label-wrapper"
  >
    <label
      className="label__lg"
      htmlFor="new-todo-input"
      style={
        Object {
          "fontSize": 20,
          "marginBottom": 25,
        }
      }
    >
      What needs to be done today?
    </label>
  </h2>
  <input
    autoComplete="off"
    className="input input__lg"
    id="new-todo-input"
    name="text"
    onChange={[Function]}
    style={
      Object {
        "borderRadius": "0.25rem",
      }
    }
    type="text"
    value=""
  />
  <div
    className="btn-group"
  >
    <button
      className="btn btn__primary"
      type="submit"
    >
      New task
    </button>
    <button
      className="btn btn__danger"
      onClick={[Function]}
    >
      Delete all tasks
    </button>
  </div>
</form>
`;
